subsys/net/lib/http/http_server.c

```diff
@@ -125,3 +169,1341 @@ int http_response_404(struct http_server_ctx *ctx, const char *html_payload)
+}
+
+static inline void new_server(struct http_server_ctx *ctx,
+			      const char *server_banner,
+			      const struct sockaddr *addr)
+{
+#if defined(CONFIG_NET_DEBUG_HTTP) && (CONFIG_SYS_LOG_NET_LEVEL > 2)
+#if defined(CONFIG_NET_IPV6)
+#define PORT_STR sizeof("[]:xxxxx")
+	char buf[NET_IPV6_ADDR_LEN + PORT_STR];
+#elif defined(CONFIG_NET_IPV4) && !defined(CONFIG_NET_IPV6)
+#define PORT_STR sizeof(":xxxxx")
+	char buf[NET_IPV4_ADDR_LEN + PORT_STR];
+#endif
+
+	if (addr) {
+		NET_INFO("%s %s (%p)", server_banner,
+			 sprint_ipaddr(buf, sizeof(buf), addr), ctx);
+	} else {
+		NET_INFO("%s (%p)", server_banner, ctx);
+	}
+#endif /* CONFIG_NET_DEBUG_HTTP */
+}
+
+static int on_header_field(struct http_parser *parser,
+			   const char *at, size_t length)
+{
+	struct http_server_ctx *ctx = parser->data;
+
+	if (ctx->req.field_values_ctr >= CONFIG_HTTP_HEADER_FIELD_ITEMS) {
+		return 0;
+	}
+
+	ctx->req.field_values[ctx->req.field_values_ctr].key = at;
+	ctx->req.field_values[ctx->req.field_values_ctr].key_len = length;
+
+	return 0;
+}
+
+static int on_header_value(struct http_parser *parser,
+			   const char *at, size_t length)
+{
+	struct http_server_ctx *ctx = parser->data;
+
+	if (ctx->req.field_values_ctr >= CONFIG_HTTP_HEADER_FIELD_ITEMS) {
+		return 0;
+	}
+
+	ctx->req.field_values[ctx->req.field_values_ctr].value = at;
+	ctx->req.field_values[ctx->req.field_values_ctr].value_len = length;
+
+	ctx->req.field_values_ctr++;
+
+	return 0;
+}
+
+static int on_url(struct http_parser *parser, const char *at, size_t length)
+{
+	struct http_server_ctx *ctx = parser->data;
+
+	ctx->req.url = at;
+	ctx->req.url_len = length;
+
+	return 0;
+}
+
+static int parser_init(struct http_server_ctx *ctx)
+{
+	memset(ctx->req.field_values, 0, sizeof(ctx->req.field_values));
+
+	ctx->req.settings.on_header_field = on_header_field;
+	ctx->req.settings.on_header_value = on_header_value;
+	ctx->req.settings.on_url = on_url;
+
+	http_parser_init(&ctx->req.parser, HTTP_REQUEST);
+
+	ctx->req.parser.data = ctx;
+
+	return 0;
+}
+
+static int http_url_cmp(const char *url, u16_t url_len,
+			const char *root_url, u16_t root_url_len)
+{
+	if (url_len < root_url_len) {
+		return -EINVAL;
+	}
+
+	if (memcmp(url, root_url, root_url_len) == 0) {
+		if (url_len == root_url_len) {
+			return 0;
+		}
+
+		/* Here we evaluate the following conditions:
+		 * root_url = /images, url = /images/ -> OK
+		 * root_url = /images/, url = /images/img.png -> OK
+		 * root_url = /images/, url = /images_and_docs -> ERROR
+		 */
+		if (url_len > root_url_len) {
+			if (root_url[root_url_len - 1] == '/') {
+				return 0;
+			}
+
+			if (url[root_url_len] == '/') {
+				return 0;
+			}
+		}
+	}
+
+	return -EINVAL;
+}
+
+static struct http_root_url *http_url_find(struct http_server_ctx *http_ctx)
+{
+	u16_t url_len = http_ctx->req.url_len;
+	const char *url = http_ctx->req.url;
+	struct http_root_url *root_url;
+	u8_t i;
+	int ret;
+
+	for (i = 0; i < CONFIG_HTTP_SERVER_NUM_URLS; i++) {
```

Comment:
Doing a linear search here means that the order which URLs were added to the `urls` array matters when matching.  You might want to consider either using a prefix tree (which is costly WRT to memory), or, instead of returning the first hit, storing the last match/prefix size, and only changing that if the last stored size is smaller than the current iteration.  (Or you can sort them by prefix length as well and compare the larger prefixes first.  Just have to figure out when to sort.)

This way, if you had something like "/sensors" to list all the sensors in a board, and "/sensors/temperature" to get the temperature sensor, you would never match "/sensors" if asked for a specific one.